File: exercise/ts/day16/tests/taskAssignmentSystem.spec.ts
import {TaskAssignmentSystem} from "../src/taskAssignmentSystem";
import {Elf} from "../src/elf";

describe('TaskAssignmentSystem', () => {
    let system: TaskAssignmentSystem;

    beforeEach(() => {
        const elves = [new Elf(1, 5), new Elf(2, 10), new Elf(3, 20)];
        system = new TaskAssignmentSystem(elves);
    });

    test('reportTaskCompletion increases total tasks completed', () => {
        expect(system.reportTaskCompletion(1)).toBeTruthy();
        expect(system.getTotalTasksCompleted()).toBe(1);
    });

    test('getElfWithHighestSkill returns the correct elf', () => {
        const highestSkillElf = system.getElfWithHighestSkill();
        expect(highestSkillElf?.id).toBe(3);
    });

    test('assignTask assigns an elf based on skill level', () => {
        expect(system.assignTask(8)).toStrictEqual(new Elf(2, 10));
    });

    test('increaseSkillLevel updates elf skill correctly', () => {
        system.increaseSkillLevel(1, 3);
        const elf = system.assignTask(7);
        expect(elf?.id).toBe(1);
    });

    test('decreaseSkillLevel updates elf skill correctly', () => {
        system.decreaseSkillLevel(1, 3);
        system.decreaseSkillLevel(2, 5);

        const elf = system.assignTask(4);
        expect(elf?.id).toBe(2);
        expect(elf?.skillLevel).toBe(5);
    });

    test('assignTaskBasedOnAvailability assigns an available elf', () => {
        const elf = system.assignTaskBasedOnAvailability(10);
        expect(elf).not.toBeNull();
    });

    test('reassignTask changes assignment correctly', () => {
        system.reassignTask(3, 1);
        const elf = system.assignTask(19);
        expect(elf?.id).toBe(1);
    });

    test('assignTask fails when skills required is too high', () => {
        expect(system.assignTask(50)).toBeUndefined();
    });

    test('listElvesBySkillDescending returns elves in correct order', () => {
        const sortedElves = system.listElvesBySkillDescending();
        expect(sortedElves.map(elf => elf.id)).toEqual([3, 2, 1]);
    });

    test('resetAllSkillsToBaseline resets all elves skills to a specified baseline', () => {
        system.resetAllSkillsToBaseline(10);
        const elves = system.listElvesBySkillDescending();
        elves.forEach(elf => expect(elf.skillLevel).toBe(10));
    });

    test('decreaseSkillLevel updates elf skill and do not allow negative values', () => {
        system.decreaseSkillLevel(1, 10);
        const elf = system.assignTask(4);
        expect(elf?.id).toBe(1);
        expect(elf?.skillLevel).toBe(5);
    });
});

File: exercise/scala/day16/src/test/scala/TaskAssignmentSystemSpec.scala
import org.scalatest.BeforeAndAfterEach
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers
import system.{Elf, TaskAssignment}

class TaskAssignmentSystemSpec extends AnyFunSuite with Matchers with BeforeAndAfterEach {
  var system: TaskAssignment = _

  override def beforeEach(): Unit = {
    val elves = List(Elf(1, 5), Elf(2, 10), Elf(3, 20))
    system = new TaskAssignment(elves)
  }

  test("reportTaskCompletion increases total tasks completed") {
    system.reportTaskCompletion(1) shouldBe true
    system.totalTasksCompleted() shouldBe 1
  }

  test("getElfWithHighestSkill returns the correct elf") {
    val highestSkillElf = system.elfWithHighestSkill()
    highestSkillElf.map(_.id) shouldBe Some(3)
  }

  test("assignTask assigns an elf based on skill level") {
    val elf = system.assignTask(8)
    elf should not be empty
    elf.map(_.id) shouldBe Some(2)
  }

  test("increaseSkillLevel updates elf skill correctly") {
    system.increaseSkillLevel(1, 3)
    val elf = system.assignTask(7)
    elf should not be empty
    elf.map(_.id) shouldBe Some(1)
  }

  test("decreaseSkillLevel updates elf skill correctly") {
    system.decreaseSkillLevel(1, 3)
    system.decreaseSkillLevel(2, 5)
    val elf = system.assignTask(4)
    elf should not be empty
    elf.map(_.id) shouldBe Some(2)
    elf.map(_.skillLevel) shouldBe Some(5)
  }

  test("reassignTask changes assignment correctly") {
    system.reassignTask(3, 1)
    val elf = system.assignTask(19)
    elf should not be empty
    elf.map(_.id) shouldBe Some(1)
  }

  test("assignTask fails when skills required is too high") {
    val elf = system.assignTask(50)
    elf shouldBe empty
  }

  test("listElvesBySkillDescending returns elves in correct order") {
    val sortedElves = system.listElvesBySkillDescending()
    sortedElves.map(_.id) shouldBe List(3, 2, 1)
  }

  test("resetAllSkillsToBaseline resets all elves skills to a specified baseline") {
    system.resetAllSkillsToBaseline(10)
    val elves = system.listElvesBySkillDescending()
    elves.foreach(_.skillLevel shouldBe 10)
  }

  test("decreaseSkillLevel updates elf skill and do not allow negative values") {
    system.decreaseSkillLevel(1, 10)
    val elf = system.assignTask(4)
    elf should not be empty
    elf.map(_.id) shouldBe Some(1)
    elf.map(_.skillLevel) shouldBe Some(5)
  }
}